framework/src/modules/chain/loader.js

```diff
@@ -881,37 +366,90 @@ __private.loadBlocksFromNetwork = function(cb) {
-										loadBlocksFromNetworkErr
-									);
-									return waterCb(
-										`Failed to load blocks from the network. ${loadBlocksFromNetworkErr}`
+					async function getBlocksFromNetwork() {
+						const lastBlock = modules.blocks.lastBlock;
+						// TODO: If there is an error, invoke the applyPenalty action on the Network module once it is implemented.
+						// TODO: Rename procedure to include target module name. E.g. chain:blocks
+						let data;
+						try {
+							const response = await library.channel.invoke('network:request', {
+								procedure: 'blocks',
+								data: {
+									lastBlockId: lastBlock.id,
+								},
+							});
+							data = response.data;
+						} catch (p2pError) {
+							library.logger.error(
+								'Failed to load block from network',
+								p2pError
+							);
+							return [];
+						}
+
+						if (!data) {
+							throw new Error(
+								'Received an invalid blocks response from the network'
+							);
+						}
+						// Check for strict equality for backwards compatibility reasons.
+						// The misspelled data.sucess is required to support v1 nodes.
+						// TODO: Remove the misspelled data.sucess === false condition once enough nodes have migrated to v2.
+						if (data.success === false || data.sucess === false) {
+							throw new Error(
+								`Peer did not have a matching lastBlockId. ${data.message}`
+							);
+						}
+						return data.blocks;
+					},
+					function validateBlocks(blocks, waterCb) {
+						const report = library.schema.validate(
+							blocks,
+							definitions.WSBlocksList
+						);
+
+						if (!report) {
+							return setImmediate(
+								waterCb,
+								new Error('Received invalid blocks data')
+							);
+						}
+						return setImmediate(waterCb, null, blocks);
+					},
+					async function loadBlocksFromNetwork(blocks) {
+						const lastBlock = modules.blocks.lastBlock;
+						try {
+							const lastValidBlock = await modules.blocks.loadBlocksFromNetwork(
+								blocks
+							);
+							__private.blocksToSync = lastValidBlock.height;
+							loaded = lastValidBlock.id === lastBlock.id;
+							// Reset counter after a batch of blocks was successfully loaded from the network
+							failedAttemptsToLoad = 0;
+						} catch (loadBlocksFromNetworkErr) {
+							if (modules.peers.isPoorConsensus()) {
+								library.logger.debug(
+									'Perform chain recovery due to poor consensus'
+								);
+								try {
+									await modules.blocks.recoverChain();
+								} catch (recoveryError) {
+									throw new Error(
+										`Failed chain recovery after failing to load blocks while network consensus was low. ${recoveryError}`
 									);
 								}
-								__private.blocksToSync = lastValidBlock.height;
-								loaded = lastValidBlock.id === lastBlock.id;
-								// Reset counter after a batch of blocks was successfully loaded from the network
-								failedAttemptsToLoad = 0;
-								return waterCb();
+								throw new Error(
```

Comment:
This will throw when `modules.blocks.recoverChain()` succeeds.
Is this the correct behaviour?

Reply:
It seems in either cases, it is on error path
https://github.com/LiskHQ/lisk-sdk/blob/development/framework/src/modules/chain/loader.js#L896
https://github.com/LiskHQ/lisk-sdk/blob/development/framework/src/modules/chain/loader.js#L905

Reply:
But note that [here](https://github.com/LiskHQ/lisk-sdk/blob/development/framework/src/modules/chain/loader.js#L894) it uses `return` before the function meaning the second one isn't executed.

Reply:
Yes second one is not executed, but `waterCb` is called with the first argument (error string), so I think throw is correct here